### PR TITLE
rocSPARSE examples (BSR format)

### PR DESCRIPTION
**Pull request by Beanavil**
This pull request contains the (first batch of the) new rocSPARSE examples. This first group of examples corresponds to the ones that showcase BSR-format-related functionality.

Examples added:
- level 2
  - `bsrmv`
  - `bsrsv`
  - `bsrxmv`
- level 3
  - `bsrmm`
  - `bsrsm`
- preconditioner
  - `bsric0`
  - `bsrilu0`

### CODE REVIEW

**Comment by Beanavil** [comment]
cc @YvanMokwinski

File: Common/rocsparse_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef COMMON_ROCSPARSE_UTILS_HPP
+#define COMMON_ROCSPARSE_UTILS_HPP
```
Comment (YvanMokwinski) [inline Common/rocsparse_utils.hpp:24]: it's just naming but rocsparse_example_utils might be more appropriate ?

File: Common/rocsparse_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    // switch. If the value is not a valid rocsparse_status, we return the following.
+    return "<undefined rocsparse_status value>";
+}
+
+/// \brief Checks if the provided status code is \p rocsparse_status_success and if not,
+/// prints an error message to the standard error output and terminates the program
+/// with an error code.
+#define ROCSPARSE_CHECK(condition)                                                               \
+    {                                                                                            \
+        const rocsparse_status status = condition;                                               \
```
Comment (YvanMokwinski) [inline Common/rocsparse_utils.hpp]: (condition)

**Review by YvanMokwinski — APPROVED** [review]
Very fine work, it looks very good. Please merge.
I don't have much to say, there is a minor comment about the name of common header that could be more specific but otherwise, code wise, documentation wise it all looks excellent.

It would be nice to have a tool to generate a squeleton for adding a new example. Or maybe an existing squeleton with 'intructions' to keep everything homogeneous.

File: Common/rocsparse_utils.hpp (hunk 1 above)
Comment (Beanavil) [reply Common/rocsparse_utils.hpp:24]: We just remained consistent with the util headers that we already had:
- `hipblas_utils.hpp`
- `hipsolver_utils.hpp`
- `rocblas_utils.hpp`

File: Common/rocsparse_utils.hpp (hunk 2 above)
Comment (Beanavil) [reply Common/rocsparse_utils.hpp]: Fixed!

**Review by saadrahim — APPROVED** [review]

**Comment by Beanavil** [comment]
> It would be nice to have a tool to generate a squeleton for adding a new example. Or maybe an existing squeleton with 'intructions' to keep everything homogeneous.

About this @YvanMokwinski, we used to have a template for adding examples to the repo but it was requested from AMD to remove this template a while ago (see https://github.com/amd/rocm-examples/issues/14). So I'm not sure if this should be added for rocSPARSE.

**Pull request merged**